ws-client/src/transport.rs

```diff
@@ -174,64 +176,80 @@ impl<'a> WsTransportClientBuilder<'a> {
+		self,
+		mut tls_connector: Option<crate::tokio::TlsConnector>,
 	) -> Result<(Sender, Receiver), WsHandshakeError> {
-		// Try establish the TCP connection.
-		let tcp_stream = {
-			let socket = TcpStream::connect(sockaddr);
-			let timeout = crate::tokio::sleep(self.timeout);
-			futures::pin_mut!(socket, timeout);
-			match future::select(socket, timeout).await {
-				future::Either::Left((socket, _)) => {
-					let socket = socket?;
-					if let Err(err) = socket.set_nodelay(true) {
-						log::warn!("set nodelay failed: {:?}", err);
-					}
-					match tls_connector {
-						None => TlsOrPlain::Plain(socket),
-						Some(connector) => {
-							let dns_name = crate::tokio::DNSNameRef::try_from_ascii_str(&self.target.host)?;
-							let tls_stream = connector.connect(dns_name, socket).await?;
-							TlsOrPlain::Tls(tls_stream)
+		let mut sockaddrs = self.target.sockaddrs;
+		let mut path = PathBuf::from(self.target.path);
+		let mut host = self.target.host;
+		let mut host_header = self.target.host_header;
+
+		let mut err = Err(None);
+
+		let client = loop {
+			let sockaddr = match sockaddrs.pop() {
+				Some(addr) => addr,
+				None => break err,
+			};
+
+			let tcp_stream = connect(sockaddr, self.timeout, &host, &tls_connector).await?;
+			let mut client = WsRawClient::new(
+				BufReader::new(BufWriter::new(tcp_stream)),
+				&host_header,
+				path.to_str().expect("valid UTF-8 checked by Url::parse; qed"),
+			);
+			if let Some(origin) = self.origin_header.as_ref() {
+				client.set_origin(origin);
+			}
+			// Perform the initial handshake.
+			match client.handshake().await? {
+				ServerResponse::Accepted { .. } => break Ok(client),
+				ServerResponse::Rejected { status_code } => {
+					err = Err(Some(WsHandshakeError::Rejected { status_code }));
+				}
+				ServerResponse::Redirect { status_code, location } => {
+					log::trace!("recv redirection: status_code: {}, location: {}", status_code, location);
+					match url::Url::parse(&location) {
+						// redirection with absolute path => need to lookup.
+						Ok(url) => {
+							let target = Target::parse(url)?;
+							sockaddrs = target.sockaddrs;
+							path = PathBuf::from(target.path);
+							host = target.host;
+							host_header = target.host_header;
+							tls_connector = match target.mode {
+								Mode::Tls => {
+									let mut client_config = rustls::ClientConfig::default();
+									if let CertificateStore::Native = self.certificate_store {
+										client_config.root_store = rustls_native_certs::load_native_certs()
+											.map_err(|(_, e)| WsHandshakeError::CertificateStore(e))?;
+									}
+									Some(Arc::new(client_config).into())
+								}
+								Mode::Plain => None,
+							};
 						}
-					}
+						// redirection is relative, either `/baz` or `bar`.
+						Err(_) => {
+							// replace the entire path if `location` is `/`.
+							if location.starts_with('/') {
+								path = location.into();
+							} else {
```

Comment:
This, in combination with `Err(_)` seems like it could be a bit dangerous. What if the `location` is the empty string such that we end up with the same path as we had before the redirect? Or if the [error](https://docs.rs/url/2.2.2/url/enum.ParseError.html) is `Overflow` and we append 4Gb of data here? Maybe I'm being paranoid but I would feel better if we did explicit matching on the `ParseError` variants we know are safe to use here?

Reply:
> What if the location is the empty string such that we end up with the same path as we had before the redirect? 

That must be some violation of HTTP spec to send back the same path as in the original request (but I haven't checked it...)

> Or if the error is Overflow and we append 4Gb of data here? Maybe I'm being paranoid but I would feel better if we did explicit matching on the ParseError variants we know are safe to use here?

fair enough, I agree with matching on the error and just make this assumption if the error is `RelativeUrl_x` then.

Reply:
> That must be some violation of HTTP spec to send back the same path as in the original request (but I haven't checked it...)

I am sure it is indeed a violation. And also that there are many many misconfigured redirects out there, servers ping-ponging requests in eternity...